### PR TITLE
#116 Allow token processing "middleware"

### PR DESCRIPTION
**Pull request by nsantini**
I added node-spellchecker to check for typos, and also "levenshtein" to find the closest spell correction to the original word.

I also modified the "negation" feature to look backwards until a negation word or a new afinn word is found, to cover cases like "not too bad".

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -63,12 +65,17 @@ Sentiment.prototype.analyze = function (phrase, opts, callback) {
     // Iterate over tokens
     var i = tokens.length;
     while (i--) {
-        var obj = tokens[i];
+        var obj = spelling.getSpellCheckedAfinnWord(labels, tokens[i]);
         if (!labels.hasOwnProperty(obj)) continue;
+
+        // Check for negation
```
Comment (nsantini) [inline lib/index.js]: When I originally forked the repo, here we used to check if the previous word was a negation and invert the score. Now seems like we deal with negation further down. So we end up negating the score twice. Might be that my change to deal with negative words is redundant now

**Comment by nsantini** [comment]
Ok
- moved the negation backwards strategy to the language model
- added unit test for it

**Comment by nsantini** [comment]
Finally, made spell checking optional, and unit tested it

**Comment by elyas-bhy** [comment]
Thanks for the quick update @nsantini, great work!
A few more things:

1. You seem to have added the distance/spellchecking feature exclusively to the english language (the only officially supported language for now, but other languages can be dynamically registered, see [Adding new languages](https://github.com/thisandagain/sentiment#adding-new-languages).
What would it take to make this feature accessible to all languages? Is there a way to set a locale for the spellchecker? If so, we can move the distance/spellchecking modules up a level, and make them available to all languages.

2. Could you please add the necessary documentation to the README file? Documenting the spellcheck flag would be quite helpful.

3. Tests look good to me. If you manage to get around the first point of this post, consider adding another test to make sure that your changes are also supported for a new language.

We're almost there!

**Comment by nsantini** [comment]
The current implementation uses https://www.npmjs.com/package/spellchecker , which are native bindings to NSSpellChecker, Hunspell, or the Windows 8 Spell Check API, depending on your platform. Windows 7 and below as well as Linux will rely on Hunspell. So it depends on what language you have in your system. Also it depends on those subsystems to support multiple languages.
So, in a way, the spellchecking implementation supports multiple languages.
I added the method signature to the language definition, so other language could support different implementations of the spellchecker.
So the question would be, do we want all language spellcheckers to be implemented in one way depending on the used library, or allow each language to use what they want?

**Comment by elyas-bhy** [comment]
Does this mean that, for example, if my OS is set to use the spanish locale, then the spellchecker module would always try to spellcheck with spanish as the target language, regardless of the actual language of the input? If this is the case, I feel like this would lead to a very inconsistent behaviour.

Is there a way to specify to the spellchecker which language to use?
/cc @thisandagain, thoughts?

**Comment by nsantini** [comment]
There is a mechanism to add a dictionary to the spellchecker, I believe thats how a language gets "set"

**Comment by nsantini** [comment]
I have refactored the spell checking functionality to use a different library [nspell](https://www.npmjs.com/package/nspell) that can support multiple languages.
Also updated README with examples of how to use spell checking

**Comment by elyas-bhy** [comment]
This is all great work, we're almost there!

1. If we officially add support to one or more new languages, we would have to duplicate the code for the spellchecking feature (since it is currently located in `languages/en`).
Do you think you can move the spellchecking / distance logic up a couple of levels and make it part of the core module (`lib`)? This would make the feature available to all languages.
At the language-level, I think it is best if we only provide the appropriate dictionary, and delegate the rest of the processing to the core module. It would make adding new languages much easier.

For example:

```js
// languages/fr/index.js
module.exports = {
    labels: require('./labels.json'),
    dictionary: require('dictionary-fr'),  // specify the language dictionary here for spellchecking
    scoringStrategy: require('./scoring-strategy')
};
```
---

2. I have noticed a drastic performance hit when comparing before/after this PR:
Before:
```
sentiment (Latest) - Short  x 561,830 ops/sec ±1.76% (92 runs sampled)
sentiment (Latest) - Long   x 2,689 ops/sec ±1.41% (87 runs sampled)
Sentimental (1.0.1) - Short x 314,373 ops/sec ±1.50% (89 runs sampled)
Sentimental (1.0.1) - Long  x 1,171 ops/sec ±2.00% (88 runs sampled)
```

After:
```
sentiment (Latest) - Short  x 239,996 ops/sec ±1.64% (91 runs sampled)
sentiment (Latest) - Long   x 0.21 ops/sec ±3.55% (5 runs sampled)
Sentimental (1.0.1) - Short x 236,762 ops/sec ±5.24% (82 runs sampled)
Sentimental (1.0.1) - Long  x 895 ops/sec ±2.54% (80 runs sampled)
```

Ideally there should be barely any impact when the spellchecking feature is disabled (which it should be, by default). Could you please investigate this issue?

---

3. I have also noticed a slight decrease in code coverage (`npm run test:coverage`). Could you add a few more tests to cover those cases?

---

4. And finally, do you mind documenting this feature in the API Reference section of the README file as well?

---

Thank you for your patience @nsantini and for bearing with me!

**Comment by thisandagain** [comment]
Looking good @nsantini! This is great to see. Agreed with @elyas-bhy's suggestions above.

**Comment by thisandagain** [comment]
In addition to the performance implications of this change necessitating the feature be "off" by default, I think the validation tests also strongly suggest that this should be optional:

### Before
```
Amazon accuracy: 0.7202797202797203
IMDB accuracy: 0.7642357642357642
Yelp accuracy: 0.6943056943056943
```

### After
```
Amazon accuracy: 0.7302697302697303
IMDB accuracy: 0.7532467532467533
Yelp accuracy: 0.6943056943056943
```

Nice improvement (~1%) on the Amazon validation set, but the IMDB accuracy dropped ~1.1% and Yelp stayed stable. This may suggest that the change is indeed helping with less formal speech, but is actually causing false positives in a more formal corpus, but more investigation would be required.

File: languages/en/distance.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+var lev = require('levenshtein');
+var tokenize = require('../../lib/tokenize');
+
+/**
+ * Finds the closest match between a statement and a body of words using
+ * Levenshtein Distance
```
Comment (thisandagain) [inline languages/en/distance.js]: Levenshtein Distance is a great performant strategy, but if we are going to make this optional anyway we might want to discuss alternative edit distance algorithms (e.g. Myers Diff Algorithm).

PDF:
[myers.pdf](https://github.com/thisandagain/sentiment/files/2115441/myers.pdf)

**Comment by nsantini** [comment]
Ok, moved the spell checking to the library and left loading the dictionary to the language. Added extra unit test, only two lines on the spell checking are not covered, but they are there for safekeeping of an edge case that shouldnt happen.
Looking into performance and accuracy:

sentiment (Latest) - Short  x 493,203 ops/sec ±1.14% (90 runs sampled)
sentiment (Latest) - Long   x 1,452 ops/sec ±3.65% (82 runs sampled)
Sentimental (1.0.1) - Short x 306,074 ops/sec ±2.98% (86 runs sampled)
Sentimental (1.0.1) - Long  x 1,283 ops/sec ±2.45% (87 runs sampled)

Amazon accuracy: 0.7302697302697303
IMDB accuracy: 0.7532467532467533
Yelp accuracy: 0.6943056943056943

The changes to use spell checking, off by default, should not have affected accuracy, since I haven't changed how is tested.

File: languages/en/dictionary.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+var read = require('fs').readFileSync;
+var dictiaonary = null;
+
+module.exports = function() {
+    if (dictiaonary ===  null) {
+        var base = require.resolve('dictionary-en-us');
+        dictiaonary = {
+            'aff': read(base.replace('.js', '.aff'), 'utf-8'),
```
Comment (elyas-bhy) [inline languages/en/dictionary.js]: Could you explain why do we need to perform such gymnastics when loading the dictionary, instead of simply requiring it and passing it over to `nspell`, [as described in their usage example](https://www.npmjs.com/package/nspell#usage)?

Also, there is a typo in the spelling of the "dictionary" variable.

File: README.md [hunk 4]
```diff
@@ -63,6 +73,12 @@ var frLanguage = {
       }
       return tokenScore;
     }
+  },
+  getDictionary: {
+    apply: function() {
+      // Load a dictionary for the language for nspell to use
+      return { aff, dic };
```
Comment (elyas-bhy) [inline README.md]: What do these `aff` and `dic` properties correspond to? Could you give an example?

Comment (nsantini) [reply README.md]: Explained more in the README file

File: languages/en/dictionary.js (hunk 3 above)
Comment (nsantini) [reply languages/en/dictionary.js]: Fixed the typo.

Regarding "gymnastics": the dictionary can be loaded on an async fashion. But the whole sentiment library works synchronously. So I decided to not pollute the whole library with callbacks or promises, and given that I couldn't use async/await to do it (eslint complained about it) I decided to load it this way.

File: README.md (hunk 4 above)
Comment (elyas-bhy) [reply README.md]: This doesn't seem to be the case, are you sure that you have pushed your changes?

**Comment by nsantini** [comment]
Hi, latest changes:
- README references the documentation of `nspell` and how to add dictionaries

File: README.md (hunk 4 above)
Comment (nsantini) [reply README.md]: new commit, it references the explanation I gave in the "Spell checked example" section

**Comment by elyas-bhy** [comment]
LGTM. @thisandagain?

**Comment by nsantini** [comment]
@thisandagain any more comments on this?

Regarding your comment about looking for alternative string distance algorithms, I would suggest leave this PR as it is (since the current implementation is performant and correct), and open a new issue to look for alternatives

**Comment by nsantini** [comment]
@elyas-bhy @thisandagain any updates?

**Comment by elyas-bhy** [comment]
Everything looks good on my side. Still waiting for @thisandagain to approve the changes.

**Comment by nsantini** [comment]
hi @thisandagain , just checking if you got a change to review the changes

**Comment by elyas-bhy** [comment]
Ping @thisandagain.

**Comment by nsantini** [comment]
Should I close this PR?
@thisandagain @elyas-bhy

**Comment by elyas-bhy** [comment]
I'm really looking forward to get this merged, but need @thisandagain to approve and merge this PR.

**Comment by elyas-bhy** [comment]
Ping @thisandagain
